v2/api/appconfiguration/customizations/configuration_store_extensions.go

```diff
@@ -0,0 +1,149 @@
+ * Licensed under the MIT license.
+ */
+
+package customizations
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/appconfiguration/armappconfiguration"
+	storage "github.com/Azure/azure-service-operator/v2/api/appconfiguration/v1beta20220501storage"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/extensions"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/secrets"
+)
+
+var _ extensions.SecretsRetriever = &ConfigurationStoreExtension{}
+
+func (ext *ConfigurationStoreExtension) RetrieveSecrets(
+	ctx context.Context,
+	obj genruntime.ARMMetaObject,
+	armClient *genericarmclient.GenericClient,
+	log logr.Logger) ([]*v1.Secret, error) {
+
+	// This has to be the current hub storage version. It will need to be updated
+	// if the hub storage version changes.
+	typedObj, ok := obj.(*storage.ConfigurationStore)
+	if !ok {
+		return nil, errors.Errorf("cannot run on unknown resource type %T, expected *appconfiguration.ConfigurationStore", obj)
+	}
+
+	// Type assert that we are the hub type. This will fail to compile if
+	// the hub type has been changed but this extension has not
+	var _ conversion.Hub = typedObj
+
+	hasSecrets := secretsSpecified(typedObj)
+	if !hasSecrets {
+		log.V(Debug).Info("No secrets retrieval to perform as operatorSpec is empty")
+		return nil, nil
+	}
+
+	id, err := genruntime.GetAndParseResourceID(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	keys := make(map[string]armappconfiguration.APIKey)
+	// Only bother calling ListKeys if there are secrets to retrieve
+	if hasSecrets {
+		subscription := armClient.SubscriptionID()
+		// Using armClient.ClientOptions() here ensures we share the same HTTP connection, so this is not opening a new
+		// connection each time through
+		var confClient *armappconfiguration.ConfigurationStoresClient
+		confClient, err = armappconfiguration.NewConfigurationStoresClient(subscription, armClient.Creds(), armClient.ClientOptions())
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create new ConfigurationStoresClient")
+		}
+
+		var pager *runtime.Pager[armappconfiguration.ConfigurationStoresClientListKeysResponse]
+		var resp armappconfiguration.ConfigurationStoresClientListKeysResponse
+		pager = confClient.NewListKeysPager(id.ResourceGroupName, obj.AzureName(), nil)
+		resp, err = pager.NextPage(ctx)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to retreive response")
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed listing keys")
+		}
+
+		keys = secretsByName(resp.Value)
```

Comment:
You need a loop here to iterate through all the pages - there's no guarantee that everything you want will be available in the first page, nor even that the page will contain anything. AFAIK, it is entirely within contract to return empty pages or single row pages.

Reply:
Done!